dace/frontend/fortran/ast_components.py

```diff
@@ -0,0 +1,1034 @@
+        children = self.create_children(node)
+        name = get_child(children, ast_internal_classes.Name_Node)
+        args = get_child(children, ast_internal_classes.Arg_List_Node)
+        return ast_internal_classes.Subroutine_Stmt_Node(name=name, args=args.args, line_number=node.item.span)
+
+    def ac_value_list(self, node: FASTNode):
+        children = self.create_children(node)
+        return ast_internal_classes.Ac_Value_List_Node(value_list=children)
+
+    def power_expr(self, node: FASTNode):
+        children = self.create_children(node)
+        line = get_line(node)
+        #child 0 is the base, child 2 is the exponent
+        #child 1 is "**"
+        return ast_internal_classes.Call_Expr_Node(name=ast_internal_classes.Name_Node(name="pow"),
+                                                   args=[children[0], children[2]],
+                                                   line_number=line)
+
+    def array_constructor(self, node: FASTNode):
+        children = self.create_children(node)
+        value_list = get_child(children, ast_internal_classes.Ac_Value_List_Node)
+        return ast_internal_classes.Array_Constructor_Node(value_list=value_list.value_list)
+
+    def structure_constructor(self, node: FASTNode):
+        children = self.create_children(node)
+        name = get_child(children, ast_internal_classes.Type_Name_Node)
+        args = get_child(children, ast_internal_classes.Component_Spec_List_Node)
+        return ast_internal_classes.Structure_Constructor_Node(name=name, args=args.args, type=None)
+
+    def intrinsic_name(self, node: FASTNode):
+        name = node.string
+        replacements = {
+            "INT": "__dace_int",
+            "DBLE": "__dace_dble",
+            "SQRT": "sqrt",
+            "COSH": "cosh",
+            "ABS": "abs",
+            "MIN": "min",
+            "MAX": "max",
+            "EXP": "exp",
+            "EPSILON": "__dace_epsilon",
+            "TANH": "tanh",
+            "SUM": "__dace_sum",
+            "SIGN": "__dace_sign",
+            "EXP": "exp",
+            "SELECTED_INT_KIND": "__dace_selected_int_kind",
+            "SELECTED_REAL_KIND": "__dace_selected_real_kind",
+        }
+        return ast_internal_classes.Name_Node(name=replacements[name])
+
+    def intrinsic_function_reference(self, node: FASTNode):
+        children = self.create_children(node)
+        line = get_line(node)
+        name = get_child(children, ast_internal_classes.Name_Node)
+        args = get_child(children, ast_internal_classes.Arg_List_Node)
+        if name.name == "__dace_selected_int_kind":
+            import math
+            return ast_internal_classes.Int_Literal_Node(value=str(
+                math.ceil((math.log2(math.pow(10, int(args.args[0].value))) + 1) / 8)),
+                                                         line_number=line)
+        # TODO This needs a better translation
+        elif name.name == "__dace_selected_real_kind":
```

Comment:
This function needs documentation - I have no idea what is happening there. It seems to hardcode some important Fortran magic.

Reply:
Couldn't this be stored as a mapping `Dict[ Tuple[str, str], str]`? The key contains string arguments of if conditions, whereas the value should be the parameter of `Int_Literal_Node`. This is a suggestion to improve readability.